File: trunk/Geometry/curvature2D.cpp
#ifndef TWODCURVATURE_H_
#define TWODCURVATURE_H_

#include <cmath>
#include <map>
#include <new>
using namespace std;

#include "geoquant.h"
#include "triposition.h"
#include "geoquants.h"
#include "triangulation/triangulation.h"

Curvature2DIndex* Curvature2D::Index = NULL;

Curvature2D::Curvature2D( Vertex& v ){    
    angles = new vector<GeoQuant*>();

    GeoQuant* angle;
    for(int i = 0; i < v.getLocalFaces()->size(); i++) {
        angle = EuclideanAngle::At(v, Triangulation::faceTable[(*(v.getLocalFaces()))[i]]);
	    angle->addDependent(this);
	    angles->push_back( angle );
    }
}

void Curvature2D::remove() {
     deleteDependents();
     for(int ii = 0; ii < angles->size(); ii++){
       angles->at(ii)->removeDependent(this);
     }    
     Index->erase(pos);
     delete angles;
}

Curvature2D::~Curvature2D(){ remove(); }
void Curvature2D::recalculate() {
    double curv = 2*PI;
    GeoQuant* angle;
    for(int ii = 0; ii < angles->size(); ii++){
      angle = angles->at(ii);
      curv -= angle->getValue();
    }
    
    value = curv;
}

Curvature2D* Curvature2D::At( Vertex& v  ){
  TriPosition T( 1, v.getSerialNumber() );
  if( Index == NULL ) Index = new Curvature2DIndex();
  Curvature2DIndex::iterator iter = Index->find( T );

  if( iter == Index->end() ){
    Curvature2D* val = new Curvature2D( v );
    val->pos = T;
    Index->insert( make_pair( T, val ) );
    return val;
  } else {
    return iter->second;
  }
}

void Curvature2D::CleanUp(){
  if( Index == NULL ) return;
  Curvature2DIndex::iterator iter;
  Curvature2DIndex copy = *Index;
  for(iter = copy.begin(); iter != copy.end(); iter++) {
    delete iter->second;
  }
    
  delete Index;
  Index = NULL;
}

#endif /* TWODCURVATURE_H_ */

File: SKUNK/src/Geoquant/Curvature2D.java
package Geoquant;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;

import Triangulation.Edge;
import Triangulation.Face;
import Triangulation.Vertex;

public class Curvature2D extends Geoquant {
  // Index map
  private static HashMap<TriPosition, Curvature2D> Index = new HashMap<TriPosition, Curvature2D>();
  
  // Needed geoquants
  private LinkedList<Angle> angles;
  
  public Curvature2D(Vertex v) {
    super();
    
    angles = new LinkedList<Angle>();
    Angle a;
    for(Face f : v.getLocalFaces()) {
      a = Angle.At(v, f);
      a.addDependent(this);
      angles.add(a);
    }
  }
  
  protected void recalculate() {
    value = 2 * Math.PI;

    for(Angle a : angles) {
      value -= a.getValue();
    }
  }
 
  protected void remove() {
    deleteDependents();
    for(Angle a : angles) {
      a.removeDependent(this);
    }
    angles.clear();
    Index.remove(pos);
  }
  
  public static Curvature2D At(Vertex v) {
    TriPosition T = new TriPosition(v.getSerialNumber());
    Curvature2D q = Index.get(T);
    if(q == null) {
      q = new Curvature2D(v);
      q.pos = T;
      Index.put(T, q);
    }
    return q;
  }
  
  public static double valueAt(Vertex v) {
    return At(v).getValue();
  }

}
